### ISSUE TITLE
Unsealing data despite PCR policy

### ISSUE THREAD

**Issue by havolarkis** [tpm2-software/tpm2-tools#1123]
Folowing the examples in branch 3.X test scripts 'test/system/test_tpm2_unseal.sh' for sealing I've got some undesirable behavior. It is possible to unseal the secret data wihtout passing the PCR list and still get it. 

```
# tpm2_pcrlist -Q -L sha1:0,1,2,3 -o pcr.bin
# tpm2_createpolicy -Q -P -L sha1:0,1,2,3 -F pcr.bin -f policy.data
# tpm2_create -Q -g sha256 -G keyedhash -u opu_sha256_keyedhash -r opr_sha256_keyedhash -I- -c context.p_sha256_ecc -L policy.data -A 'sign|fixedtpm|fixedparent|sensitivedataorigin'
# tpm2_load -Q -c context.p_sha256_ecc -u opu_sha256_keyedhash -r opr_sha256_keyedhash -n name.load_sha256_ecc-sha256_keyedhash -C ctx_load_out_sha256_ecc-sha256_keyedhash
# tpm2_unseal -c ctx_load_out_sha256_ecc-sha256_keyedhash -L sha1:0,1,2,3 -F pcr.bin
12345678
# tpm2_pcrextend 3:sha1=6c10289a8da7f774cf67bd2fc8502cd4b585346a
# tpm2_unseal -c ctx_load_out_sha256_ecc-sha256_keyedhash
12345678
# tpm2_unseal -c ctx_load_out_sha256_ecc-sha256_keyedhash -L sha1:0,1,2,3 -F pcr.bin
ERROR: Failed parse_policy_type_and_send_command
ERROR: Building PCR policy failed: 0x1c4
```

If I do not pass the option `-L sha1:0,1,2,3 -F pcr.bin`, I still get the secret sealed. Am I missing something? Is it some sort of TPM2 feature?

I am on a CentOS-7.5
ibmtpm1119.tar.gz (tpm server)
tpm2-abrmd-1.1.0
tpm2-tools-3.0.1
tpm2-tss-1.3.0

**Comment by diabonas** [comment]
If you remove the `-Q` option from the `tpm2_create` command, you will see that the key is created with the `userwithauth` attribute. This means that instead of unlocking the secret based on the PCR configuration, you can also unlock it with the password specified by the `-p` parameter to `tpm2_create`. If you do not set a password, it will be left empty, so anybody can unlock the secret. This is exactly what is happening in your case: as you did not specify a PCR list in the `tpm2_unseal` call, it gets unlocked by the default empty password instead.

To fix this, replace

    tpm2_create -Q -g sha256 -G keyedhash -u opu_sha256_keyedhash -r opr_sha256_keyedhash -I- -c context.p_sha256_ecc -L policy.data -A 'sign|fixedtpm|fixedparent|sensitivedataorigin'

by

    tpm2_create -Q -g sha256 -G keyedhash -u opu_sha256_keyedhash -r opr_sha256_keyedhash -I- -c context.p_sha256_ecc -L policy.data -A 0x492

as suggested in https://github.com/tpm2-software/tpm2-tools/issues/80#issuecomment-232518202. Unfortunately, the more human-readable

    tpm2_create -Q -g sha256 -G keyedhash -u opu_sha256_keyedhash -r opr_sha256_keyedhash -I- -c context.p_sha256_ecc -L policy.data -A 'fixedtpm|fixedparent|adminwithpolicy|noda'

does *not* work as `userwithauth` is still added to the attributes list due to #1097 and 352908f7470338aba858a9ac20c1c5ec608b661c.

**Comment by joshuagl** [comment]
It might be useful to document this in some kind of FAQ on the wiki.

@havolarkis FWIW I can't reproduce the error in the final line of your example with a more recent stack: tpm2-tss 2.0, tpm2-abrmd 2.0 and tpm2-tools 3.1.x -- is upgrading an option?

**Comment by havolarkis** [comment]
@diabonas thanks, that was exactly was I needed. 

@joshuagl Upgrade just wouldn't be the easiest way. But for now the tpm2-tools-3.0.x undestands the hex code of `userwithauth`.  So this was my bad for not trying. 

I agree that might be nice to document some basic tpm2 operations, like sealing/unsealing process. If I can do any help on this, just ask.

**Comment by williamcroberts** [comment]
@havolarkis we love wiki contributions.

**Comment by diabonas** [comment]
@havolarkis No problem, you are welcome! The `userwithauth` attribute was something I was struggling with myself when I started using this tool suite...

@joshuagl I get the same error with tpm2-tools 3.1.2-rc0, tpm2-tss 2.0.0 and tpm2-2.0.1: in my setup, that is the standard error I get when the file with the expected PCR values does not match the actual PCR state.

**Comment by martinezjavier** [comment]
@havolarkis you closed the issue but IIUC it hasn't been solved yet (at least for 3.0.4 and 3.1.2).  I've commented my thoughts on this in issue #1097.
